core/trino-main/src/main/java/io/trino/metadata/QualifiedFunctionName.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.metadata;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class QualifiedFunctionName
+{
+    private final Optional<String> catalogName;
+    private final Optional<String> schemaName;
+    private final String functionName;
+
+    public static QualifiedFunctionName of(String functionName)
+    {
+        return new QualifiedFunctionName(Optional.empty(), Optional.empty(), functionName);
+    }
+
+    public static QualifiedFunctionName of(String schemaName, String functionName)
+    {
+        return new QualifiedFunctionName(Optional.empty(), Optional.of(schemaName), functionName);
+    }
+
+    public static QualifiedFunctionName of(String catalogName, String schemaName, String functionName)
+    {
+        return new QualifiedFunctionName(Optional.of(catalogName), Optional.of(schemaName), functionName);
+    }
+
+    private QualifiedFunctionName(Optional<String> catalogName, Optional<String> schemaName, String functionName)
+    {
+        this.catalogName = requireNonNull(catalogName, "catalogName is null");
```

Comment:
`requireNonNull` not needed since `catalogName` is dereferenced below (the JVM will produce a better error message in that case)